prometheus/histogram.go

```diff
@@ -1835,3 +1840,208 @@ func (n *nativeExemplars) addExemplar(e *dto.Exemplar) {
 		n.exemplars = append(n.exemplars[:nIdx], append([]*dto.Exemplar{e}, append(n.exemplars[nIdx:rIdx], n.exemplars[rIdx+1:]...)...)...)
 	}
 }
+
+type constNativeHistogram struct {
+	desc *Desc
+	dto.Histogram
+	labelPairs []*dto.LabelPair
+}
+
+func validateCount(sum float64, count uint64, negativeBuckets, positiveBuckets map[int]int64, zeroBucket uint64) error {
+	var bucketPopulationSum int64
+	for _, v := range positiveBuckets {
+		bucketPopulationSum += v
+	}
+	for _, v := range negativeBuckets {
+		bucketPopulationSum += v
+	}
+	bucketPopulationSum += int64(zeroBucket)
+
+	// If the sum of observations is NaN, the number of observations must be greater or equal to the sum of all bucket counts.
+	// Otherwise, the number of observations must be equal to the sum of all bucket counts .
+
+	if math.IsNaN(sum) && bucketPopulationSum > int64(count) ||
+		!math.IsNaN(sum) && bucketPopulationSum != int64(count) {
+		return fmt.Errorf("the sum of all bucket populations exceeds the count of observations")
+	}
+	return nil
+}
+
+func validateExemplars(exemplars []*dto.Exemplar) error {
+	for _, e := range exemplars {
+		if !e.GetTimestamp().IsValid() {
+			return fmt.Errorf("invalid timestamp found in exemplars")
+		}
+	}
+	return nil
+}
+
+// NewConstNativeHistogram returns a metric representing a Prometheus Nativehistogram with
+// fixed values for the count, sum, and positive/negative/zero bucket counts. As those parameters
+// cannot be changed, the returned value does not implement the Histogram
+// interface (but only the Metric interface). Users of this package will not
+// have much use for it in regular operations. However, when implementing custom
+// Collectors, it is useful as a throw-away metric that is generated on the fly
+// to send it to Prometheus in the Collect method.
+//
+// zeroBucket counts all (positive and negative)
+// observations in the zero bucket (with an absolute value less or equal
+// the current threshold)
+// positiveBuckets and negativeBuckets are separate maps for negative and positive
+// observations. The map's value is an int64, counting observations in
+// that bucket. The map's key is the
+// index of the bucket according to the used
+// Schema. Index 0 is for an upper bound of 1.
+// NewConstNativeHistogram returns an error if
+// 1.the length of labelValues is not consistent with the variable labels in Desc or if Desc is invalid.
+// 2.the schema passed is not between 8 and -4
+// 3.the sum of counts in the bucket do not equal count
+func NewConstNativeHistogram(
+	desc *Desc,
+	count uint64,
+	sum float64,
+	positiveBuckets, negativeBuckets map[int]int64,
+	zeroBucket uint64,
+	schema int32,
+	zeroThreshold float64,
+	createdTimestamp time.Time,
+	exemplars []*dto.Exemplar,
```

Comment:
We shouldn't use naked `dto.Exemplar` here. Check out how it is done for `NewMetricWithExemplars`, i.e. there is a `prometheus.Exemplar` and a `newExemplar` constructor to create the `dto.Exemplar`.

Looking at `NewMetricWithExemplars`, however, I'm wondering if we should even allow an exemplar argument here. We could also implement native histogram handling in `withExemplarsMetric` (strictly speaking, we have to do that anyway), and then all `NewConst...` functions would be consistent (as in: not having exemplars, but you use the `NewMetricWithExemplars` to wrap them).

@ArthurSens @bwplotka WDYT?

Reply:
I'll wait for a consensus here.

Reply:
Sounds reasonable to me :), let's keep this pattern.

Reply:
To unblock this PR and make it simpler, maybe here we could leave exemplars out. We deal with exemplars in a follow-up PR that implements `.withExemplars(e []*dto.Exemplar)`. Would that work for both of you @beorn7 @shivanthzen?

Reply:
Yes, that's essentially what I sketched up as plan B above. As said, we have to make `NewMetricWithExemplars` work with native histograms anyway, and then we don't really need it derictly in `NewConstNativeHistogram` anymore.